Disallow subclassing when a class is missing vtable entries.

jrose-apple opened the pull request:
Builds on top of #9360—only the last commit is different. This isn't an inherent limitation of the language—in fact, it would be a problem for library evolution if you had to know a superclass's full vtable contents to generate the vtable for a subclass. However, that's exactly where we are today, and that's not going to change for Swift 4.

One small hole in the Swift 3 / Swift 4 story.

More rdar://problem/31878396

jrose-apple committed:
Disallow subclassing when a class is missing vtable entries.

This isn't an inherent limitation of the language---in fact, it would
be a problem for library evolution if you had to know a superclass's
full vtable contents to generate the vtable for a subclass. However,
that's exactly where we are today, and that's not going to change for
Swift 4.

One small hole in the Swift 3 / Swift 4 story.

More rdar://problem/31878396

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
@swift-ci Please smoke test

jckarter commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1908,6 +1908,10 @@ ERROR(inheritance_from_final_class,none,
       "type parameters of %0 to specific concrete types", (Identifier))
+ERROR(inheritance_from_class_with_missing_vtable_entries,none,
+      "cannot inherit from class %0 in Swift %1 mode because it has "
+      "overridable members that cannot be loaded",
```
Can the diagnostic be phrased in a more user-comprehensible way? "Cannot be loaded" isn't going to mean much to most people, and sounds like a condition that could be the user's fault. Can we describe the reason why the member couldn't be loaded (SDK change between language versions, etc.)?

jrose-apple replied:
By this point that information has been thrown away, and even within deserialization it's hard to say. ("I can't load this because I couldn't load the thing it overrides because I couldn't find this particular type, probably because you changed language versions.") I can at least preserve the language version change, though.

jckarter replied:
Since version mix and match is the most likely cause of deserialization recovery for Swift 4, maybe having a specialized diagnostic for that case is good enough.

jrose-apple committed:
Specifically mention version mismatches for non-subclassable classes.

That is, if a Swift 4 class has members that cannot be loaded in Swift
3 mode and that's why it can't be subclassed, mention that in the
diagnostic.

jrose-apple committed:
Mention version mismatches for non-implementable protocols as well.

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
@jckarter Any further comments?

jckarter commented:
LGTM. I still wonder if there's still a better phrase to use than "could not be loaded in <version>".

jrose-apple commented:
I think the most useful thing would be to tell users which members couldn't be loaded, even if they don't know why. I'll add that in a follow-up commit.
